docker.qmd

```diff
@@ -0,0 +1,49 @@
+
+## Docker Image Guidelines
+
+- Because these Docker images will be used for individual steps within WDL workflows, they should be as minimal as possible in terms of the number of tools installed in each image (1 or 2 max).
+- As a general (but flexible) rule, try to start from as basic of a parent image as possible, e.g. `scratch`, `ubuntu`, `python`, `r-base`, etc.
+    - Outside parent images are fine, as long as they are from a VERY trusted source, e.g. Ubuntu, Python, Conda, Rocker, etc.
+- To speed up build and deployment of containers, try to keep image sizes relatively small (a few hundred MB on average, 2GB max).
+    - For this reason, reference data should not be stored in an image unless absolutely necessary.
+    - Unnecessary tools should also be avoided, even if they serve a "just-in-case" functionality.
+
+## Dockerfile Guidelines
+- Every Dockerfile must contain the labels below at a minimum. This provides users with increased visibility in terms of where the image came from and open access to the necessary resources in case they have any questions or concerns.
+```
+LABEL org.opencontainers.image.title="awesomeimage" # Name of the image in question
+LABEL org.opencontainers.image.description="Short description of awesomeimage and its purpose"
+LABEL org.opencontainers.image.version="1.0" # Version tag of the image
+LABEL org.opencontainers.image.authors="johndoe@fredhutch.org" # Author email address
+LABEL org.opencontainers.image.url=https://hutchdatascience.org/ # Home page
+LABEL org.opencontainers.image.documentation=https://getwilds.org/ # Documentation page
+LABEL org.opencontainers.image.source=https://github.com/getwilds/wilds-docker-library # GitHub repo to link with
+LABEL org.opencontainers.image.licenses=MIT # License type for the image in question
+```
+- When creating a different version of an existing image, use one of the other Dockerfiles as a starting template and modify it as needed.
+    - This will help to ensure that the only thing that has changed between image versions is the version of tool in question, not any strange formatting/configuration issues.
+- Try to be as specific as possible in terms of tool versions within the Dockerfile, especially the parent image.
+    - If you just specify "latest", a tag that get updated frequently over time, your image could be completely different the next time you build it, even though it uses the exact same Dockerfile.
+    - On the other hand, specifying "v1.2.3" will always pull the same instance of the tool every time, providing greater reproducibility over time.
+
+## Repository Guidelines
+
+- In terms of the repo organization, each image should have its own directory named after the tool being used in the image. Each version of the image should have its own Dockerfile in that directory following the naming convention of `[IMAGENAME]/Dockerfile_[VERSIONTAG]`.
```

Comment:
This is a lot more work, but I think that having examples might be really helpful to some of the more abstract guidelines. At least with relevant lines, and you can point towards code examples in getwilds.

Reply:
Great call, I think links to code examples in getWILDS would be perfect, especially for the WDL guidelines.